src/Features/Core/Portable/ConvertForEachToFor/AbstractConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,474 @@
+        {
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                return;
+            }
+
+            // TODO: refactor introduce variable refactoring to real service and use that service here to introduce local variable
+            var generator = editor.Generator;
+
+            // attach rename annotation to control variable
+            var collectionVariableToken = generator.Identifier(collectionVariable.ToString()).WithAdditionalAnnotations(RenameAnnotation.Create());
+
+            // this expression is from user code. don't simplify this.
+            var expression = foreachCollectionExpression.WithoutAnnotations(SimplificationHelpers.DontSimplifyAnnotation);
+            var collectionStatement = generator.LocalDeclarationStatement(
+                type,
+                collectionVariableToken,
+                foreachInfo.RequireExplicitCastInterface
+                    ? generator.CastExpression(foreachInfo.ExplicitCastInterface, expression) : expression);
+
+            // attach trivia to right place
+            collectionStatement = collectionStatement.WithLeadingTrivia(foreachInfo.ForEachStatement.GetFirstToken().LeadingTrivia);
+
+            editor.InsertBefore(foreachInfo.ForEachStatement, collectionStatement);
+        }
+
+        protected SyntaxNode AddItemVariableDeclaration(
+            SyntaxGenerator generator, SyntaxNode type, SyntaxToken foreachVariable,
+            ITypeSymbol castType, SyntaxNode collectionVariable, SyntaxToken indexVariable)
+        {
+            var memberAccess = generator.ElementAccessExpression(
+                    collectionVariable, generator.IdentifierName(indexVariable));
+
+            return generator.LocalDeclarationStatement(
+                type, foreachVariable, generator.CastExpression(castType, memberAccess));
+        }
+
+        private ForEachInfo GetForeachInfo(
+            ISemanticFactsService semanticFact, OptionSet options, SemanticModel model,
+            TForEachStatement foreachStatement, CancellationToken cancellationToken)
+        {
+            var operation = model.GetOperation(foreachStatement, cancellationToken) as IForEachLoopOperation;
+            if (operation == null || operation.Locals.Length != 1)
+            {
+                return null;
+            }
+
+            var foreachVariable = operation.Locals[0];
+            if (foreachVariable == null)
+            {
+                return null;
+            }
+
+            // VB can have Next variable. but we only support
+            // simple 1 variable case.
+            if (operation.NextVariables.Length > 1)
+            {
+                return null;
+            }
+
+            if (!operation.NextVariables.IsEmpty)
+            {
+                var nextVariable = operation.NextVariables[0] as ILocalReferenceOperation;
+                if (nextVariable == null || nextVariable.Local?.Equals(foreachVariable) == false)
```

Comment:
shouldn't this be != true....    or is it intentional that this will succeed if Local is null?  if so, this warrants comments.